File: 6_kyu/Detect_Pangram.cs
public static class Kata {
    // Array of integers, value = 2**index
    private static readonly int[] bits = {
        1,
        2,
        4,
        8,
        16,
        32,
        64,
        128,
        256,
        512,
        1024,
        2048,
        4096,
        8192,
        16384,
        32768,
        65536,
        131072,
        262144,
        524288,
        1048576,
        2097152,
        4194304,
        8388608,
        16777216,
        33554432
    };

    // Returns True if str contains all letters a..z, is a pangram.
    public static bool IsPangram(string str) {
        if (str.Length < 26) return false;
        str = str.ToLower();
        int b = 0;
        foreach(char c in str) if ('a' <= c && c <= 'z') b |= bits[c - 'a'];
        return b == 67108863;
    }
}

File: 6_kyu/Detect_Pangram.rs
// Number of letters
const LETTER_A: usize = 'A' as usize;
const LETTER_Z: usize = 'Z' as usize;
const LETTERS_N: usize = 1 + LETTER_Z - LETTER_A;

// Checks if string s is a pangram.
// Checks if string s contains all letters a..z.
fn is_pangram(s: &str) -> bool {
    if s.len() < LETTERS_N {
        // The string is too short to hold all letters.
        return false;
    }
    // number of not seen letters
    let mut not_seen = LETTERS_N;
    // seen letters
    let mut seen = [false; LETTERS_N];
    for c in s.chars() {
        // Ignores other letters like, öäü.
        if c.is_ascii_alphabetic() {
            // The simple to_ascii_uppercase works fpr a..z.
            let index = c.to_ascii_uppercase() as usize - LETTER_A;
            if !seen[index] {
                seen[index] = true;
                not_seen -= 1;
                if not_seen == 0 {
                    // All letters seen, exit
                    return true;
                }
            }
        }
    }
    // Not seen all letters
    false
}
